File: SOLID/code-examples/chapter-09.py
# FILE: chapter-09-isp.py
# Chapter 9 — Interface Segregation Principle — Beginner/Intermediate

from abc import ABC, abstractmethod
from typing import List

# BAD EXAMPLE - Violates ISP (Fat Interface)
class BadPrinter(ABC):
    @abstractmethod
    def print_document(self, document: str) -> None:
        pass
    
    @abstractmethod
    def scan_document(self, document: str) -> None:  # Not all printers can scan
        pass
    
    @abstractmethod
    def fax_document(self, document: str) -> None:   # Not all printers can fax
        pass
    
    @abstractmethod
    def email_document(self, document: str) -> None: # Not all printers can email
        pass

class SimplePrinter(BadPrinter):
    def print_document(self, document: str) -> None:
        print(f"Printing: {document}")
    
    # Forced to implement methods it doesn't support - ISP violation!
    def scan_document(self, document: str) -> None:
        raise NotImplementedError("Simple printer cannot scan")
    
    def fax_document(self, document: str) -> None:
        raise NotImplementedError("Simple printer cannot fax")
    
    def email_document(self, document: str) -> None:
        raise NotImplementedError("Simple printer cannot email")

# GOOD EXAMPLE - Follows ISP (Segregated Interfaces)

# Separate interfaces for different capabilities
class Printable(ABC):
    @abstractmethod
    def print_document(self, document: str) -> None:
        pass

class Scannable(ABC):
    @abstractmethod
    def scan_document(self, document: str) -> None:
        pass

class Faxable(ABC):
    @abstractmethod
    def fax_document(self, document: str) -> None:
        pass

class Emailable(ABC):
    @abstractmethod
    def email_document(self, document: str) -> None:
        pass

# Simple printer only implements what it can do
class GoodSimplePrinter(Printable):
    def print_document(self, document: str) -> None:
        print(f"Simple printer printing: {document}")

# Multifunction printer implements multiple interfaces
class MultiFunctionPrinter(Printable, Scannable, Faxable, Emailable):
    def print_document(self, document: str) -> None:
        print(f"MFP printing: {document}")
    
    def scan_document(self, document: str) -> None:
        print(f"MFP scanning: {document}")
    
    def fax_document(self, document: str) -> None:
        print(f"MFP faxing: {document}")
    
    def email_document(self, document: str) -> None:
        print(f"MFP emailing: {document}")

# Scanner that only scans and emails
class NetworkScanner(Scannable, Emailable):
    def scan_document(self, document: str) -> None:
        print(f"Network scanner scanning: {document}")
    
    def email_document(self, document: str) -> None:
        print(f"Network scanner emailing: {document}")

# Office class that uses different printer capabilities
class Office:
    def __init__(self):
        self.printers: List[Printable] = []
        self.scanners: List[Scannable] = []
    
    def add_printer(self, printer: Printable) -> None:
        self.printers.append(printer)
    
    def add_scanner(self, scanner: Scannable) -> None:
        self.scanners.append(scanner)
    
    def print_documents(self, document: str) -> None:
        print(f"\nPrinting '{document}' on all available printers:")
        for printer in self.printers:
            printer.print_document(document)
    
    def scan_documents(self, document: str) -> None:
        print(f"\nScanning '{document}' on all available scanners:")
        for scanner in self.scanners:
            scanner.scan_document(document)

def main():
    print("=== ISP Violation Demo ===")
    
    bad_printer = SimplePrinter()
    bad_printer.print_document("Report.pdf")  # Works fine
    
    try:
        bad_printer.scan_document("Invoice.pdf")  # Will raise exception - ISP violation!
    except NotImplementedError as e:
        print(f"Error: {e}")

    print("\n=== ISP Compliant Demo ===")
    
    # Demonstrate ISP benefits
    office = Office()
    
    # Create different devices
    simple_printer = GoodSimplePrinter()
    mfp = MultiFunctionPrinter()
    network_scanner = NetworkScanner()
    
    # Add devices based on their capabilities
    office.add_printer(simple_printer)  # Can only print
    office.add_printer(mfp)            # Can print (among other things)
    
    office.add_scanner(mfp)            # Can scan (among other things)
    office.add_scanner(network_scanner) # Can scan (among other things)
    
    # Use devices through segregated interfaces
    office.print_documents("Monthly Report")
    office.scan_documents("Invoice")
    
    print("\n=== ISP Benefits ===")
    print("✓ Simple printer only implements print - no unused methods")
    print("✓ Office can use any printer without knowing its other capabilities")
    print("✓ Easy to add new device types with different capability combinations")
    print("✓ No risk of calling unsupported methods")
    print("✓ Each interface is focused and cohesive")
    
    # Demonstrate flexibility - can treat MFP as different device types
    print("\n=== Flexibility Demo ===")
    just_printer: Printable = mfp  # MFP used only as printer
    just_scanner: Scannable = mfp  # Same MFP used only as scanner
    
    just_printer.print_document("Using MFP as just a printer")
    just_scanner.scan_document("Using MFP as just a scanner")

if __name__ == "__main__":
    main()

File: SOLID/code-examples/chapter-09.cpp
// FILE: chapter-09-isp.cpp
// Chapter 9 — Interface Segregation Principle — Beginner/Intermediate

#include <iostream>
#include <vector>
#include <memory>
#include <stdexcept>

// BAD EXAMPLE - Violates ISP (Fat Interface)
class BadPrinter {
public:
    virtual ~BadPrinter() = default;
    virtual void print(const std::string& document) = 0;
    virtual void scan(const std::string& document) = 0;    // Not all printers can scan
    virtual void fax(const std::string& document) = 0;     // Not all printers can fax
    virtual void email(const std::string& document) = 0;   // Not all printers can email
};

class SimplePrinter : public BadPrinter {
public:
    void print(const std::string& document) override {
        std::cout << "Printing: " << document << std::endl;
    }
    
    // Forced to implement methods it doesn't support - ISP violation!
    void scan(const std::string& document) override {
        throw std::runtime_error("Simple printer cannot scan");
    }
    
    void fax(const std::string& document) override {
        throw std::runtime_error("Simple printer cannot fax");
    }
    
    void email(const std::string& document) override {
        throw std::runtime_error("Simple printer cannot email");
    }
};

// GOOD EXAMPLE - Follows ISP (Segregated Interfaces)

// Separate interfaces for different capabilities
class Printable {
public:
    virtual ~Printable() = default;
    virtual void print(const std::string& document) = 0;
};

class Scannable {
public:
    virtual ~Scannable() = default;
    virtual void scan(const std::string& document) = 0;
};

class Faxable {
public:
    virtual ~Faxable() = default;
    virtual void fax(const std::string& document) = 0;
};

class Emailable {
public:
    virtual ~Emailable() = default;
    virtual void email(const std::string& document) = 0;
};

// Simple printer only implements what it can do
class GoodSimplePrinter : public Printable {
public:
    void print(const std::string& document) override {
        std::cout << "Simple printer printing: " << document << std::endl;
    }
};

// Multifunction printer implements multiple interfaces
class MultiFunctionPrinter : public Printable, public Scannable, public Faxable, public Emailable {
public:
    void print(const std::string& document) override {
        std::cout << "MFP printing: " << document << std::endl;
    }
    
    void scan(const std::string& document) override {
        std::cout << "MFP scanning: " << document << std::endl;
    }
    
    void fax(const std::string& document) override {
        std::cout << "MFP faxing: " << document << std::endl;
    }
    
    void email(const std::string& document) override {
        std::cout << "MFP emailing: " << document << std::endl;
    }
};

// Scanner that only scans and emails
class NetworkScanner : public Scannable, public Emailable {
public:
    void scan(const std::string& document) override {
        std::cout << "Network scanner scanning: " << document << std::endl;
    }
    
    void email(const std::string& document) override {
        std::cout << "Network scanner emailing: " << document << std::endl;
    }
};

// Office class that uses different printer capabilities
class Office {
private:
    std::vector<std::unique_ptr<Printable>> printers;
    std::vector<std::unique_ptr<Scannable>> scanners;

public:
    void addPrinter(std::unique_ptr<Printable> printer) {
        printers.push_back(std::move(printer));
    }
    
    void addScanner(std::unique_ptr<Scannable> scanner) {
        scanners.push_back(std::move(scanner));
    }
    
    void printDocuments(const std::string& document) {
        std::cout << "\nPrinting '" << document << "' on all available printers:" << std::endl;
        for (const auto& printer : printers) {
            printer->print(document);
        }
    }
    
    void scanDocuments(const std::string& document) {
        std::cout << "\nScanning '" << document << "' on all available scanners:" << std::endl;
        for (const auto& scanner : scanners) {
            scanner->scan(document);
        }
    }
};

int main() {
    std::cout << "=== ISP Violation Demo ===" << std::endl;
    
    SimplePrinter badPrinter;
    badPrinter.print("Report.pdf");  // Works fine
    
    try {
        badPrinter.scan("Invoice.pdf");  // Will throw exception - ISP violation!
    } catch (const std::exception& e) {
        std::cout << "Error: " << e.what() << std::endl;
    }

    std::cout << "\n=== ISP Compliant Demo ===" << std::endl;
    
    // Create different devices
    auto simplePrinter = std::make_unique<GoodSimplePrinter>();
    auto mfp1 = std::make_unique<MultiFunctionPrinter>();
    auto mfp2 = std::make_unique<MultiFunctionPrinter>();
    auto networkScanner = std::make_unique<NetworkScanner>();
    
    // Demonstrate ISP benefits
    Office office;
    
    // Add devices based on their capabilities
    office.addPrinter(std::move(simplePrinter));  // Can only print
    office.addPrinter(std::move(mfp1));          // Can print (among other things)
    
    office.addScanner(std::move(mfp2));          // Can scan (among other things)
    office.addScanner(std::move(networkScanner)); // Can scan (among other things)
    
    // Use devices through segregated interfaces
    office.printDocuments("Monthly Report");
    office.scanDocuments("Invoice");
    
    std::cout << "\n=== ISP Benefits ===" << std::endl;
    std::cout << "✓ Simple printer only implements print - no unused methods" << std::endl;
    std::cout << "✓ Office can use any printer without knowing its other capabilities" << std::endl;
    std::cout << "✓ Easy to add new device types with different capability combinations" << std::endl;
    std::cout << "✓ No risk of calling unsupported methods" << std::endl;
    std::cout << "✓ Each interface is focused and cohesive" << std::endl;
    
    // Demonstrate flexibility with references
    std::cout << "\n=== Flexibility Demo ===" << std::endl;
    MultiFunctionPrinter flexibleMFP;
    
    // Same MFP object used through different interfaces
    Printable& justPrinter = flexibleMFP;   // MFP used only as printer
    Scannable& justScanner = flexibleMFP;   // Same MFP used only as scanner
    
    justPrinter.print("Using MFP as just a printer");
    justScanner.scan("Using MFP as just a scanner");
    
    return 0;
}